### PR TITLE
ignite-8780 File I/O operations must be retried if buffer hasn't read/written completely

### PR DESCRIPTION
**Pull request by SpiderRus**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/AbstractFileIO.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
+package org.apache.ignite.internal.processors.cache.persistence.file;
```
Comment (Jokser) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/AbstractFileIO.java]: Missed license header

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/AbstractFileIO.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+     */
+    private int avail(int requested) throws IOException {
+        return avail(requested, position());
+    }
+
+    /**
+     * @param requested Requested.
+     * @param position Position.
+     */
+    private int avail(int requested, long position) throws IOException {
```
Comment (Jokser) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/AbstractFileIO.java]: Please rename both method to available

File: modules/direct-io/src/test/java/org/apache/ignite/internal/processors/cache/persistence/file/IgniteFileIOTest.java [hunk 3]
```diff
@@ -0,0 +1,287 @@
+package org.apache.ignite.internal.processors.cache.persistence.file;
```
Comment (Jokser) [inline modules/direct-io/src/test/java/org/apache/ignite/internal/processors/cache/persistence/file/IgniteFileIOTest.java]: Missed license header

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/AbstractFileIO.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+                if (n > 0) {
+                    i += n;
+                    time = 0;
+                }
+                else if (n == 0) {
+                    if (time == 0)
+                        time = U.currentTimeMillis();
+                    else if ((U.currentTimeMillis() - time) >= MAX_IO_TIMEOUT_MS)
+                        throw new IOException(write && position() == size() ? "Failed to extend file." :
```
Comment (Jokser) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/AbstractFileIO.java]: Change message for this exception. Failed to extend failed was in WAL. This method is generic)

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/FileWriteAheadLogManager.java [hunk 5]
```diff
@@ -1402,29 +1402,19 @@ private void formatFile(File file, int bytesCntToFormat) throws StorageException
...
 
                 fileIO.force();
             }
             else
                 fileIO.clear();
         }
         catch (IOException e) {
-            throw new StorageException("Failed to format WAL segment file: " + file.getAbsolutePath(), e);
+            StorageException ex = new StorageException("Failed to format WAL segment file: " + file.getAbsolutePath(), e);
+            if (failureProcessor != null)
```
Comment (Jokser) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/FileWriteAheadLogManager.java]: Add Blank line

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/FileWriteAheadLogManager.java [hunk 6]
```diff
@@ -1402,29 +1402,19 @@ private void formatFile(File file, int bytesCntToFormat) throws StorageException
...
+                failureProcessor.process(new FailureContext(FailureType.CRITICAL_ERROR, ex));
+            throw ex;
```
Comment (Jokser) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/wal/FileWriteAheadLogManager.java]: Add blank line

**Comment by dspavlov** [comment]
The ticket is resolved, closing PR. Feel free to reopen if needed.

**Pull request closed without merging**